luaui/Widgets/unit_dgun_stall_assist.lua

```diff
@@ -1,59 +1,88 @@
-
 function widget:GetInfo()
 	return {
-		name      = "DGun Stall Assist",
+		name      = "DGun Stall Assist v2",
 		desc      = "Waits cons/facs when trying to dgun and stalling",
-		author    = "Niobium",
-		date      = "2 April 2010",
+		author    = "zombean",
+		date      = "2 April 2024",
 		license   = "GNU GPL, v2 or later",
 		layer     = 0,
-		enabled   = true  --  loaded by default?
+		enabled   = true
 	}
 end
 
-----------------------------------------------------------------
--- Config
-----------------------------------------------------------------
-local targetEnergy = 600
-local watchForTime = 5
+local millis_start_timer = Spring.GetTimer()
+local function millis()
+  return math.ceil(Spring.DiffTimers(Spring.GetTimer(), millis_start_timer) * 1000)
+end
 
-----------------------------------------------------------------
--- Globals
-----------------------------------------------------------------
-local watchTime = 0
-local waitedUnits = nil -- nil / waitedUnits[1..n] = uID
-local shouldWait = {} -- shouldWait[uDefID] = true / nil
-local isFactory = {} -- isFactory[uDefID] = true / nil
+local target_energy = 1000
+local dgun_cost = 500
+local time_in_ms_befor_unwaiting = 250 --how many fast we can unwait units after waiting them
+local gameStarted = false
+local DGUN_OUT = false
+local UNITS_WAITING = false
+local waitableUnits = {}
+local waitingUnits = {}
+
+local function getWaitAbleUnits()
+  local myWaitableUnits = {}
+  local myUnits = Spring.GetTeamUnits(Spring.GetMyTeamID())
+  for i=1, #myUnits do
+    local unitCmd = Spring.GetUnitCurrentCommand(myUnits[i])
+    local isEnergyConsumingCommand = not (unitCmd == CMD.WAIT or unitCmd == CMD.MOVE or unitCmd == CMD.RECLAIM)
+    if isEnergyConsumingCommand then
+      local unitDefId = Spring.GetUnitDefID(myUnits[i])
+      if waitableUnits[unitDefId] then myWaitableUnits[#myWaitableUnits + 1] = myUnits[i] end
+    end
+  end
+  return myWaitableUnits
+end
```

Comment:
This can be rewritten. There are a few issues here
- Units that are not consuming energy will get paused
- Rezzing is not identified
- It does more iteration than it needs to

This is how I've rewritten it:
```lua
local function getEnergyConsumingBuilders()
	local needsWait = {}
	local myUnits = Spring.GetTeamUnitsSorted(Spring.GetMyTeamID())
	for unitDefID, units in pairs(myUnits) do
		if waitableUnits[unitDefID] then
			for i, unit in pairs(units) do
				local metalMake, metalUse, energyMake, energyUse = spGetUnitResources(unit)
				if energyUse > 0 then
					needsWait[#needsWait + 1] = unit
				end
			end
		end
	end
	return needsWait
end
```
- `GetTeamUnitsSorted` will return an array of arrays, where each unitdef is its own array. We can completely skip any analysis of unitDefID's that we don't care about
- `GetUnitResources` will let us skip any order analysis and just check if the units are draining energy. There's a *theoretical* chance that this can give false-positives during hard stalls, however during my testing, units still try to consume a lot of energy even during hard stalls. I consider it good enough. We also use a cached version here since it will be called many times per frame.